rfcs/0051-hard-fork-package-generation.md

```diff
@@ -0,0 +1,148 @@
+
+There is a requirement from Mina Foundation that the hard fork package can be prepared in 6 hours or less, so all of these steps need to fit within that timeframe.
+
+<!-- THIS TURNED OUT TO ACTUALLY BE IMPLEMENTED
+### Required Updates to the Genesis Ledger Runtime Configuration
+
+Today, the genesis ledger runtime configuration supports specifying genesis ledgers via their contents (list of account states in the ledger). This data can be very slow to load, and even if we have an on-disk ledger to load instead, we need to hash all of the accounts in order to determine what the correct merkle root of the genesis ledger should be. Additionally, even just parsing this amount of JSON adds quite a bit of delay to initializing a daemon.
+
+To better support the workflow of having pre-packaged on-disk ledgers available to load for initial genesis ledger states, we will need to update the runtime configuration to support specifying ledgers via their merkle root instead of their set of accounts. The runtime configuration should continue to support both workflows, as the workflow of specifying specific accounts is very useful for development and testing, it is just undesirable when shipping a hard fork release.
+
+Specifically, we can update the `
+-->
+
+### Capturing Genesis State for Planned Hard Forks
+
+Under the current hard fork plan, a soft fork update will be shipped to the prior chain which bakes in shutdown logic at specific slot heights. At some slot height, the prior network will stop accepting transactions, and then at a subsequent slot height after that, the network will shutdown entirely (will stop producing/accepting/broadcasting blocks). At this point in time, we take the strongest chain, and within that chain, take the final block before the slot height we stopped accepting transactions at. This block is where we will dump the genesis state from.
+
+Rather than waiting to dump the genesis block data at the network halt slot, we will proactively dump every candidate hard fork genesis state, and select the correct candidate hard fork genesis state manually once the network halt slot is reached. There will be a special CLI flag that enables dumping the candidate hard fork genesis block states to disk automatically, which will be enabled on some nodes connected to the network. To ensure that we capture this data and do not lose it, a cron job will be executed on the nodes dumping this state which will attempt to upload any dumped candidate hard fork genesis block states to the cloud.
```

Comment:
Should we be concerned about the disk usage from dumping all of the ledgers? I don't know the actual numbers, but if we assume 1KB per account and 500 ledgers dumped, that's over 80GB.

Reply:
It's certainly something that needs to be considered in the infrastructure configuration. I can call that out explicitly.

Reply:
@Sventimir I don't think you did the dump correctly. Dumping the best tip staged ledger from a synced node on mainnet produces a JSON ledger of size `104M`. Are you sure your node was synced when you dumped it?